### DISCUSSION TITLE
Does SignUp Using Email and Password Return User Data Even for Existing Emails?

### DISCUSSION THREAD

**Discussion by jungjay-cb** [supabase/supabase#26899] [Questions] [answered]
When executing the signup method using an email and password, the request succeeds regardless of whether the email is already registered or not, and an email to confirm signup is automatically sent.

Additionally, even if the password used in the signup request is different from the previously registered one, the request still succeeds, and the existing user's data, including the auth id, is returned.

Is this the intended behavior?

Below is the code executed in the development environment (using supabase/ssr with supabase.js. in server side components)

The const supabase is defined using createServerClient from supabase/ssr.

```javascript
const credential = {
    email: input.email,
    password: input.password
};

const { data, error } = await supabase.auth.signUp(
    {
        ...credential,
        options: {
            emailRedirectTo: `${baseUrl}/auth/callback`,
        }
    }
);

console.log(data);
```

Result of console.log(data):

```
{
  user: {
    id: "5f112d9e-9c9c-49f9-a975-7b41140222dc",
    aud: "authenticated",
    role: "authenticated",
    email: "abc@abc.com",
    phone: "",
    confirmation_sent_at: "2024-05-30T22:13:48.691423103Z",
    app_metadata: { provider: "email", providers: [Array] },
    user_metadata: {
      email: "abc@abc.com",
      email_verified: false,
      phone_verified: false,
      sub: "5f112d9e-9c9c-49f9-a975-7b41140222dc"
    },
    identities: [ [Object] ],
    created_at: "2024-05-30T22:11:38.248537Z",
    updated_at: "2024-05-30T22:13:48.696557Z",
    is_anonymous: false
  },
  session: null
}
```

This shows that, regardless of whether the password is correct or not, repeated signup requests for abc@abc.com return the same ID. Checking the auth.users table in the database, it seems that the actual table ID of the user abc@abc.com is being returned.

Since the ANON KEY is exposed to the client, can anyone acquire a specific user's UUID by randomly inputting email addresses in their client or server?

Before sending a confirmation email, do we have to verify if the user already exists by other means before calling signup?

**Comment by GaryAustin1** [comment]
Are you sure it is not because you used the client to signup a real user before this and their info is still in the local storage/cookies?
I just did signUp on an email that I have not used in my browser and it returned fake info as expected.
Real UUID:
![image](https://github.com/supabase/supabase/assets/54564956/73c10761-017b-4a6b-a8a4-f3e5c02c13f9)
Fake UUID from signUp of existing email:
![image](https://github.com/supabase/supabase/assets/54564956/5a9866f9-a500-4fc0-bada-f9030a087828)

**Comment by jungjay-cb** [comment]
@GaryAustin1  Thank you for answer and I apologize for the insufficient explanation.

For users with email_verified = true, a fake UUID is returned, and the confirmation email is not sent. Thank you for the information.

I cleared all cookies and verified again, but for users with email_verified = false, when the signup method is used instead of the resend method, the confirmation email is still sent. The UUID returned appears to be the actual UUID.

While the UUID is not highly relevant to my use case, I am looking for a way to prevent the confirmation email from being sent via the signup method for users who have already been created through the signup method but have not yet confirmed their email.

Executed SignUp -> Deleted all cookies -> Executed SignUp with the same email and a different password.

As a result, the email verification was successful, but the password was the one from the initial SignUp. I thought that if a user forgot they had already executed SignUp and tried to sign up with a new password, they might get confused because the password would be different from the one they just entered.

I would like to avoid any confusion that may arise when a different password is used for signup than the one previously entered.

How can I check if a user is already registered before sending the email?

**Reply by jungjay-cb** [reply]
Or is there a security issue with checking whether an email has already been signed up? Is it more secure to keep this information hidden?

**Reply by GaryAustin1** [reply, marked as answer]
If you have "confirm email" disabled then it will just return the user right then.  Really not a good thing to ever use IMO.
If you have "confirm email" enabled then it will return a fake user if they exists.  There will not be an error as Supabase does not want to allow probing for emails.  

I really have not studied the password impacts on this if any.

https://supabase.com/docs/reference/javascript/auth-signup

Has the details such as they are.

Some users create a postgres function to check for the email with an rpc call.  This of course then opens you up to the see if a user I know is using this site security issue.  They can't use the site, but they would know that email is in use there.

**Reply by jungjay-cb** [reply]
@GaryAustin1 I understand. Thank you very much for your response. I will be trying RPC with securely.
